File: src/lexer/token.c
#include <stdio.h>

#include "lexer/token.h"

char* const TOKEN_TYPE_STRINGS[] = {
    "EOF",
    "INVALID",
    "SL_COMMENT",
    "ML_COMMENT",
};

ASSERT_ENUM_TO_STR(TOKEN_TYPE_STRINGS, TT_MAX);

char* token_type_to_string(TokenType token) {
    return TOKEN_TYPE_STRINGS[token];
}

void print_token(Token token) {
    printf("[%-15s] ~ %-60.*s ", token_type_to_string(token.type), (int)token.value_len, token.value);
    print_location(token.location);
}

File: src/lexer/token.rs
use super::location::Location;
use std::fmt::{self, Display, Formatter};

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TokenType {
    // Special tokens
    Eof,
    Invalid,

    // Single-character tokens

    // Multi-character tokens

    // Keywords

    // Types

    // Literals
    Integer,
    Float,

    // Identifiers
    Identifier,
}

impl TokenType {
    fn to_string(&self) -> String {
        format!("{:?}", self).to_uppercase()
    }
}

pub struct Token {
    tt: TokenType,
    lexeme: String,
    location: Location,
}

impl Token {
    pub fn new(tt: TokenType, lexeme: String, location: Location) -> Self {
        Self {
            tt,
            lexeme,
            location,
        }
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[{:<15}] ~ {:<60} {}",
            self.tt.to_string(),
            self.lexeme,
            self.location
        )
    }
}
